lib/src/main/java/com/diffplug/spotless/npm/NpmServerBasedFormatterStep.java

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright 2024 DiffPlug
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless.npm;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.ObjectStreamException;
+import java.io.Serializable;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.TimeoutException;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import com.diffplug.spotless.FormatterStep;
+import com.diffplug.spotless.ProcessRunner.LongRunningProcess;
+import com.diffplug.spotless.ThrowingEx;
+
+abstract class NpmServerBasedFormatterStep implements FormatterStep {
+
+	// dynamic equals/hascode impl
+	// => equals/hashcode should not include absolute paths and such
+	// => serialized/deserialized state can include absolute paths and such and should recreate a valid/runnable state
+
+	private final Map<NpmConfigElement, Serializable> configElements = new HashMap<>();
```

Comment:
This map contains all state for the FormatterStep that should be respected for equals/hashCode. Everything else is implemented as derived.